File: http/http-server-annotation-processor/src/test/java/ru/tinkoff/kora/http/server/annotation/processor/controller/TestControllerWithDifferentTypes.java
package ru.tinkoff.kora.http.server.annotation.processor.controller;

import org.reactivestreams.Publisher;
import reactor.core.publisher.Mono;
import reactor.util.function.Tuple2;
import ru.tinkoff.kora.http.common.HttpHeaders;
import ru.tinkoff.kora.http.common.annotation.HttpRoute;
import ru.tinkoff.kora.http.server.common.HttpServerRequest;
import ru.tinkoff.kora.http.server.common.HttpServerResponse;
import ru.tinkoff.kora.http.server.common.annotation.HttpController;

import java.nio.ByteBuffer;
import java.time.Duration;
import java.util.List;

import static java.nio.charset.StandardCharsets.UTF_8;
import static ru.tinkoff.kora.http.common.HttpMethod.*;

@HttpController
public class TestControllerWithDifferentTypes {
    @HttpRoute(method = GET, path = "/")
    Mono<HttpServerResponse> getRoot() {
        var response = HttpServerResponse.of(200, "text/plain", UTF_8.encode("Hello world"));

        return Mono.delay(Duration.ofMillis(1)).thenReturn(response);
    }

    @HttpRoute(method = GET, path = "/somePage")
    Mono<HttpServerResponse> getSomePage(HttpServerRequest httpServerRequest) {
        var response = HttpServerResponse.of(200, "text/plain", UTF_8.encode("Hello world"));

        return Mono.delay(Duration.ofMillis(1)).thenReturn(response);
    }

    @HttpRoute(method = POST, path = "/someEntity")
    Mono<SomeEntity> postSomeEntityById(SomeEntity someEntity) {

        return Mono.delay(Duration.ofMillis(1)).thenReturn(someEntity);
    }

    @HttpRoute(method = PUT, path = "/someEntityBlocking")
    SomeEntity postSomeEntity(SomeEntity someEntity) {
        return someEntity;
    }

    @HttpRoute(method = PATCH, path = "/someComplexGenericEntity")
    SomeEntity postSomeEntities(List<List<Tuple2<List<String>, List<SomeEntity>>>> someEntity) {
        return someEntity.get(0).get(0).getT2().get(0);
    }

    @HttpRoute(method = DELETE, path = "/deleteByteArrayVoidResult")
    public void deleteByteArrayVoidResult(byte[] data) {

    }

    @HttpRoute(method = DELETE, path = "/deleteByteArrayMonoVoidResult")
    public Mono<Void> deleteByteArrayMonoVoidResult(byte[] data) {
        return Mono.empty();
    }

    @HttpRoute(method = GET, path = "/getWithPrimitiveInt")
    int getWithPrimitiveInt(int data) {
        return data;
    }

    @HttpRoute(method = OPTIONS, path = "/publisherVoid")
    Mono<Void> postSomeFilePublisherResult(byte[] data) {
        return Mono.empty();
    }

    @HttpRoute(method = POST, path = "/publisherByteBuffer")
    Mono<Void> postSomeFilePublisherResult(Publisher<ByteBuffer> data) {
        return Mono.empty();
    }


}

File: http/http-server-symbol-processor/src/test/kotlin/ru/tinkoff/kora/http/server/symbol/processor/controllers/TestControllerWithDifferentTypes.kt
package ru.tinkoff.kora.http.server.symbol.processor.controllers

import kotlinx.coroutines.delay
import org.reactivestreams.Publisher
import reactor.util.function.Tuple2
import ru.tinkoff.kora.http.common.HttpHeaders
import ru.tinkoff.kora.http.common.HttpMethod
import ru.tinkoff.kora.http.common.annotation.HttpRoute
import ru.tinkoff.kora.http.server.common.HttpServerRequest
import ru.tinkoff.kora.http.server.common.HttpServerResponse
import ru.tinkoff.kora.http.server.common.SimpleHttpServerResponse
import ru.tinkoff.kora.http.server.common.annotation.HttpController
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

@HttpController
open class TestControllerWithDifferentTypes {
    @HttpRoute(method = HttpMethod.GET, path = "/")
    open suspend fun getRoot(): SimpleHttpServerResponse {
        val response = SimpleHttpServerResponse(
            200,
            "text/plain",
            HttpHeaders.of(),
            StandardCharsets.UTF_8.encode("Hello world")
        )
        delay(1)
        return response
    }

    @HttpRoute(method = HttpMethod.GET, path = "/somePage")
    open suspend fun getSomePage(httpServerRequest: HttpServerRequest?): HttpServerResponse {
        val response = SimpleHttpServerResponse(200, "text/plain", HttpHeaders.of(), StandardCharsets.UTF_8.encode("Hello world"))
        delay(1000)

        return  response
    }

    @HttpRoute(method = HttpMethod.POST, path = "/someEntity")
    open suspend fun postSomeEntityById(someEntity: SomeEntity): SomeEntity {
        delay(1)
        return someEntity
    }

    @HttpRoute(method = HttpMethod.PUT, path = "/someEntityBlocking")
    open fun postSomeEntity(someEntity: SomeEntity): SomeEntity {
        return someEntity
    }

    @HttpRoute(method = HttpMethod.PATCH, path = "/someComplexGenericEntity")
    open fun postSomeEntities(someEntity: List<List<Tuple2<List<String?>?, List<SomeEntity?>>>>): SomeEntity? {
        return someEntity[0][0].t2[0]
    }

    @HttpRoute(method = HttpMethod.DELETE, path = "/deleteByteArrayVoidResult")
    fun deleteByteArrayVoidResult(data: ByteArray?) {
    }

    @HttpRoute(method = HttpMethod.DELETE, path = "/deleteByteArrayMonoVoidResult")
    open suspend fun deleteByteArrayMonoVoidResult(data: ByteArray?) {
        delay(1)
    }

    @HttpRoute(method = HttpMethod.GET, path = "/getWithPrimitiveInt")
    open fun getWithPrimitiveInt(data: Int): Int {
        return data
    }

    @HttpRoute(method = HttpMethod.OPTIONS, path = "/publisherVoid")
    open suspend fun postSomeFilePublisherResult(data: ByteArray?) {
        delay(100)
    }

    @HttpRoute(method = HttpMethod.POST, path = "/publisherByteBuffer")
    open suspend fun postSomeFilePublisherResult(data: Publisher<ByteBuffer?>?) {
        delay(100)
    }
}
